winch/codegen/src/masm.rs

```diff
@@ -57,6 +57,14 @@ pub(crate) enum CmpKind {
     GeS,
     /// Unsigned greater than or equal.
     GeU,
+    /// Nature of the comparison depends on the ISA.
+    /// In x86, whether the least significant byte is even.
+    /// Not useable in AArch64.
+    Np,
```

Comment:
I think this and the `P` flag are exposed at the wrong level. This should be in the ISA-specific code and not on a cross-platform macroassembler since the instruction hypothetically results in different behaviour depending on the ISA and is not supported on some.

What I'm thinking is it might make sense to have the x86 microassembler `setcc` instruction accept a different enum as an argument and that enum would include `P` and `NP` and then we wouldn't need to have this at the macroassembler level.

Reply:
I agree that `P` and `NP` shouldn't go here, this is an implementation detail of x64.

> What I'm thinking is it might make sense to have the x86 microassembler setcc instruction accept a different enum as an argument and that enum would include P and NP and then we wouldn't need to have this at the macroassembler level.

My preference would be to simplify this even more: given that we're not interested in leaking `P` and `NP` to MacroAssembler, and given that these two variants for setcc are not used anywhere else, I think we can introduce two other functions in the assembler that deal with the specifics of setting `P` and `NP`,: `setp` and `setnp`, and leave any generalization of this case for the future in case it's needed.

Reply:
`setp` and `setnp` would directly encode the CC, something along the lines of:

```rust
pub fn setp(&mut self, dst: Reg) {
   // Clear the dst register or bits 1 to 31 may be incorrectly set.
   // Don't use xor since it updates the status register.
   self.emit(Inst::Imm {
     dst_size: args::OperandSize::Size32, // Always going to be an i32 result.
     simm64: 0,
     dst: dst.into(),
   });
        
   self.emit(Inst::Setcc {
     cc: CC::P,
     dst: dst.into(),
   });
 }
```

Reply:
I feel like we should probably extract the common logic between the three methods into a private method that takes a `CC` argument since the logic is perfectly identical except for the value provided for `cc`. I'm just not sure what a good name would be for it.

Reply:
The disadvantage with that is that we'd probably need to either: (i) introduce a common CC representation parameter that serves a use-case that can be shared to ISA-agnostic code and with code that it's not supposed to be shared (P and NP), at least not as part of this change or (ii) introduce two enums and conversions between them. Personally I've learned that with compilers doing the simplest thing first is usually better, so I still feel that introducing a bit of repetition is perfectly fine.

Reply:
But maybe you've thought through this in a different way, one that doesn't involve any of the options mentioned above. But what I'm trying to say is that I'm fine with a bit of repetition to keep things simple, since they'll probably keep changing in the future.

Reply:
I was thinking more like:

```rust
fn setcc_private(&mut self, cc: CC, dst: Reg) {
  // Clear the dst register or bits 1 to 31 may be incorrectly set.
  // Don't use xor since it updates the status register.
  self.emit(Inst::Imm {
            dst_size: args::OperandSize::Size32, // Always going to be an i32 result.
            simm64: 0,
            dst: dst.into(),
  });
  // Copy correct bit from status register into dst register.
  self.emit(Inst::Setcc {
            cc: kind.into(),
            dst: dst.into(),
    });
}

pub fn setcc(&mut self, kind: IntCmpKind, dst: Reg) {
  self.setcc_private(kind.into(), dst);
}

pub fn setp(&mut self, dst: Reg) {
  self.setcc_private(CC::P, dst);
}

pub fn setnp(&mut self, dst: Reg) {
  self.setcc_private(CC:NP, dst);
}
```

except `setcc_private` is not a good name and I'm not sure what a good name would be.

Reply:
Oh, right, that makes sense -- let's call it `setcc_impl`? This is a pattern I've used in Winch to do this kind of thing.